### PR TITLE
KAFKA-8006: Guard calls to init and close from global processor

### PR DESCRIPTION
**Pull request by ableegoldman**
Guards against users calling init() and close() on global stores accessed via global processor

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ableegoldman** [comment]
@mjsax @guozhangwang @bbejeck @vvcephei

**Review by bbejeck — APPROVED** [review]
LGTM

**Review by guozhangwang — APPROVED** [review]
LGTM.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImplTest.java [hunk 1]
```diff
@@ -129,4 +129,14 @@ public void shouldNotAllowToSchedulePunctuationsUsingDeprecatedApi() {
     public void shouldNotAllowToSchedulePunctuations() {
         globalContext.schedule(null, null, null);
     }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void shouldNotAllowInit() {
+        globalContext.getStateStore(GLOBAL_STORE_NAME).init(null, null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImplTest.java]: This is two methods calls, thus it should not be a one liner (test would pass if `getStateStore()` would throw `UnsupportedOperationException`.

Should be:
```
@Test
    public void shouldNotAllowInit() {
        final StateStore store = globalContext.getStateStore(GLOBAL_STORE_NAME);
        try {
            store.init(null, null);
            fail("Should have thrown UnsupportedOperationException.");
        } catch(final UnsupportedOperationException expected) { }
```

Similar below.

**Review by mjsax — CHANGES_REQUESTED** [review]
Overall LGTM. One nit for the unit tests.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImplTest.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImplTest.java]: Ack

**Comment by mjsax** [comment]
Actually, you should compare the tests in `ProcessorContextImplTest` -- they are quite sophisticated and it might be worth to replicate them here?

**Comment by ableegoldman** [comment]
@mjsax Do you mean the tests in general or with regards to init & close specifically? I guess it's not clear to me what benefit the sophistication is really providing here..I felt it was better to do a straightforward test of the single functionality being tested, but am open to doing otherwise

**Comment by mjsax** [comment]
Also, there is a dependency between this PR and #6173 and #6356. Depending which one we merge first, the other need to be updated.

**Comment by bbejeck** [comment]
Java 11 failed `org.apache.kafka.streams.state.internals.RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.shouldForwardAllDbOptionsCalls`
Java 8 failed
```
kafka.api.AdminClientIntegrationTest.testMinimumRequestTimeouts
kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
```
retest this please

**Comment by mjsax** [comment]
@ableegoldman The fancy test also verifies that only `init()` and `close()` are guarded and it's written to avoid test code duplication. But you raise a fair point that the fancy test is complex code. But it's already there so it might be worth do "copy and adapt" to the global case, too. Thoughts @guozhangwang @bbejeck? If we want to go with simpler test pattern, it might be worth to rewrite the existing fancy test to simplify it?

**Review by guozhangwang — COMMENTED** [review]
I looked at ProcessorContextImplTest and I agree with @ableegoldman that it seems over-complicated, we do not need to follow its pattern and even better, simplify that one as well.

**Review by mjsax — APPROVED** [review]
Thanks for the input @guozhangwang!

Merging this now. We can do follow ups if we want to simplify existing test code.

**Pull request merged**